src/metabase/lib/convert.cljc

```diff
@@ -1,10 +1,60 @@
 (ns metabase.lib.convert
   (:require
    [clojure.set :as set]
+   [malli.core :as mc]
    [medley.core :as m]
    [metabase.lib.dispatch :as lib.dispatch]
    [metabase.lib.options :as lib.options]
-   [metabase.lib.util :as lib.util]))
+   [metabase.lib.schema :as lib.schema]
+   [metabase.lib.util :as lib.util]
+   [metabase.util :as u]))
+
+(defn- clean-location [almost-stage error-type error-location]
+  (let [operate-on-parent? #{:malli.core/missing-key :malli.core/end-of-input}
+        location (if (operate-on-parent? error-type)
+                   (drop-last 2 error-location)
+                   (drop-last 1 error-location))
+        [location-key] (if (operate-on-parent? error-type)
+                         (take-last 2 error-location)
+                         (take-last 1 error-location))]
+    (if (seq location)
+      (update-in almost-stage
+                 location
+                 (fn [error-loc]
+                   (let [result (assoc error-loc location-key nil)]
+                     (cond
+                       (vector? error-loc) (into [] (remove nil?) result)
+                       (map? error-loc) (u/remove-nils result)
+                       :else result))))
+      (dissoc almost-stage location-key))))
+
+(def ^:private stage-keys-to-clean
+  #{:expressions :joins :filters :order-by :aggregation :fields :breakout})
+
+(defn- clean-stage [almost-stage]
+  (loop [almost-stage almost-stage
+         removals []]
+    (if-let [[error-type error-location] (->> (mc/explain ::lib.schema/stage.mbql almost-stage)
+                                              :errors
+                                              (filter (comp stage-keys-to-clean first :in))
+                                              (map (juxt :type :in))
+                                              first)]
+      (let [new-stage (clean-location almost-stage error-type error-location)]
+        (if (= new-stage almost-stage)
+          almost-stage
+          (recur new-stage (conj removals [error-type error-location]))))
+      almost-stage)))
```

Comment:
Just for fun: this function looks like it's iterating a function over a value until the output is the same as the input, i.e., reaching a fix point. That means it could be written like
```clojure
(->> almost-stage
     (iterate resolve-one-error)
     (partition 2 1)
     (drop-while (partial apply not=))
     ffirst)
```